[office-js, office-js-preview] ExcelAPI 1.12 Updates

AlexJerabek opened the pull request:
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [ ] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: 
https://appsforoffice.officeapps.live.com/afo/lib/1.1/hosted/office.d.ts
https://appsforoffice.officeapps.live.com/afo/lib/beta/hosted/office.d.ts
- [ ] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.
- [ ] Include [tests for your changes](https://github.com/DefinitelyTyped/DefinitelyTyped#testing)
- [ ] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`. If for reason the any rule need to be disabled, disable it for that line using `// tslint:disable-next-line [ruleName]` and not for whole package so that the need for disabling can be reviewed.

AlexJerabek commented:
Note for reviewers (including @alison-mk): This PR is based on the updates in the CDN. I made manual changes to keep TableStyles in preview, as per internal conversations. I also made manual edits for pending documentation edits and internal fields being populated here by codegen.

Rick-Kirkham commented on types/office-js-preview/index.d.ts:
```diff
@@ -43582,12 +43919,12 @@ declare namespace Excel {
     enum AutoFillType {
         /**
-         * Populates the adjacent cells based on the surrounding data (the standard AutoFill behavior).
+         * Populates the adjacent cells with data the selected data.
```
Not a sentence. I can't tell what this means.

Rick-Kirkham commented on types/office-js-preview/index.d.ts:
```diff
@@ -43582,12 +43919,12 @@ declare namespace Excel {
         fillDefault = "FillDefault",
         /**
-         * Populates the adjacent cells with data based on the selected data.
+         * Populates the adjacent cells with data the selected data.
```
Not grammatical.

Rick-Kirkham commented on types/office-js-preview/index.d.ts:
```diff
@@ -53477,24 +53868,21 @@ declare namespace Excel {
             /**
             *
-            * Returns a collection of ranges that comprises this object.
+            * Returns a collection of ranges that comprise this object.
```
See my comment on the earlier place where this same change is made.

Rick-Kirkham commented on types/office-js-preview/index.d.ts:
```diff
@@ -58941,6 +59374,70 @@ declare namespace Excel {
         interface SlicerItemCollectionData {
             items?: Excel.Interfaces.SlicerItemData[];
         }
+        /** An interface describing the data returned by calling `linkedDataType.toJSON()`. */
```
Again, why describe the interface in terms of its role in calling code? Let's describe what it is instead.

AlexJerabek replied:
This seems to be a consistent issue with all these `*Data` interfaces.

FWIW, we don't include these `*Data` classes in the TOC and the only way to see them is through a link from `*.toJSON`. Changing them will be something we'll need to work with the office-js-api team to fix. Is it worth pursuing?

Rick-Kirkham replied:
> This seems to be a consistent issue with all these `*Data` interfaces.
> 
> FWIW, we don't include these `*Data` classes in the TOC and the only way to see them is through a link from `*.toJSON`. Changing them will be something we'll need to work with the office-js-api team to fix. Is it worth pursuing?

No. Not worth pursuing.

Rick-Kirkham reviewed (changes requested):
I only looked at the preview. All my comments apply to both files. The oddest thing is that there are many descriptions for the new beta APIs that appear to be formatted incorrectly. They are indented and there are no "*" to the left of the indented lines. I marked the first few that I found, but someone needs to search all the descriptions of beta APIs and fix this.

AlexJerabek commented:
Thanks @Rick-Kirkham. I'm looking at your comments now. I will say that the formatting issues are a part of how codegen handles long descriptions. They do not appear to cause problems downstream, so I'm disinclined to manually fix them.

AlexJerabek commented on types/office-js-preview/index.d.ts:
```diff
@@ -41505,6 +41482,361 @@ declare namespace Excel {
+         */
+        getItemAt(index: number): Excel.LinkedDataType;
+        /**
+         * Gets a linked data type by id. If the linked data type does not exist, a null object is returned.
```
@alison-mk & @Rick-Kirkham - Is there an effort to fix this in the source files across the board? I'm also worried we're going to keep introducing this confusion with new versions of this method. Perhaps there's a larger fix to codegen that can be done to include this link (@wandyezj?).

Rick-Kirkham replied:
@AlexJerabek Yes. @alison-mk has an item to make this fix in all the Excel APIs in the metadata file. We should do it for the Word metadata file too.

alison-mk replied:
Ah yes, that's the file I was tracking down access to for the last few weeks. But *ta-da* I now have access! I'll plan to make the Excel metadata update after Ignite, unless y'all want me to prioritize it sooner.

Rick-Kirkham commented:
> Thanks @Rick-Kirkham. I'm looking at your comments now. I will say that the formatting issues are a part of how codegen handles long descriptions. They do not appear to cause problems downstream, so I'm disinclined to manually fix them.

OK.

AlexJerabek commented:
Thanks @Rick-Kirkham. I've addressed all of your feedback that's not part of a larger, systemic issue. I've created [an ADO work item](https://office.visualstudio.com/OC/_workitems/edit/4470611) to track those issues.

AlexJerabek commented:
Ready to merge
